src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,194 @@
+                if (symbol == null || !unassignedSymbols.Contains(symbol))
+                {
+                    continue;
+                }
+
+                if (!IsMemberOfThisInstance(descendant))
+                {
+                    unassignedSymbols.Remove(symbol);
+                }
+
+                if (IsDescendentOf<TConstructorDeclarationSyntax>(descendant, out var ctorNode))
+                {
+                    var isInAnonymousOrLocalFunction = false;
+                    for (var current = descendant.Parent; current != ctorNode; current = current.Parent)
+                    {
+                        if (syntaxFactsService.IsAnonymousOrLocalFunction(current))
+                        {
+                            isInAnonymousOrLocalFunction = true;
+                            break;
+                        }
+                    }
+
+                    if (isInAnonymousOrLocalFunction)
+                    {
+                        unassignedSymbols.Remove(symbol);
+                    }
+                    else
+                    {
+                        var ctorSymbol = model.GetDeclaredSymbol(ctorNode);
+                        if (!ctorSymbol.ContainingType.Equals(symbol.ContainingType))
+                        {
+                            unassignedSymbols.Remove(symbol);
+                        }
+
+                        if (!ctorSymbol.IsStatic && symbol.IsStatic)
+                        {
+                            unassignedSymbols.Remove(symbol);
+                        }
+                    }
+
+                    // assignments in the ctor don't matter other than the static modifiers and lambdas point checked above
+                    continue;
+                }
+                
+                if (IsWrittenTo(name, model, cancellationToken))
+                {
+                    unassignedSymbols.Remove(symbol);
+                }
+            }
+        }
+        
+        private bool IsDescendentOf<T>(SyntaxNode node, out T ctor) where T : SyntaxNode
+        {
+            ctor = node.FirstAncestorOrSelf<T>();
+            return ctor != null;
+        }
+
+        private bool IsMutableValueType(ITypeSymbol type)
+        {
+            if (type.TypeKind != TypeKind.Struct)
+            {
+                return false;
+            }
+
+            foreach (var member in type.GetMembers())
+            {
+                if (member is IFieldSymbol fieldSymbol &&
+                    !(fieldSymbol.IsConst || fieldSymbol.IsReadOnly || fieldSymbol.IsStatic))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract bool IsWrittenTo(TIdentifierNameSyntax node, SemanticModel model, CancellationToken cancellationToken);
+        protected abstract bool IsMemberOfThisInstance(SyntaxNode node);
```

Comment:
minor nit:  in general we place hte abstracts below the constructor and above the concrete methods.

Reply:
➡️ Fixed in commit 38db528